src/or/dirauth/dirvote.c

```diff
@@ -267,10 +267,13 @@ format_networkstatus_vote(crypto_pk_t *private_signing_key,
     else
       params = tor_strdup("");
     tor_assert(cert);
-    if (v3_ns->bw_file_headers)
+    if (v3_ns->bw_file_headers) {
       bw_file_headers = smartlist_join_strings(v3_ns->bw_file_headers, " ", 0,
-                                              NULL);
-    else
+                                               NULL);
+      if (smartlist_len(v3_ns->bw_file_headers) > MAX_BW_FILE_HEADERS_LEN)
+        /* FIXME */
+        smartlist_truncate(v3_ns->bw_file_headers, MAX_BW_FILE_HEADERS_LEN);
```

Comment:
We already limit the number of headers in earlier code.

So you can just log a BUG() warning here, and vote an empty string:
```
if (BUG(smartlist_len(v3_ns->bw_file_headers) > MAX_BW_FILE_HEADERS_LEN)) {
    bw_file_headers = tor_strdup("");
}
```

You should do this check before joining the string together, to avoid a memory leak, and to avoid allocating large amounts of memory.

Reply:
so then i would do 2 checks?, 1 with the smartlist and 1 with the string?.
The code then would be:
```
      if (BUG(smartlist_len(v3_ns->bw_file_headers) > MAX_BW_FILE_HEADERS_LEN)) {
        bw_file_headers = tor_strdup("");
      }
      bw_file_headers = smartlist_join_strings(v3_ns->bw_file_headers, " ", 0,
                                               NULL);
      if (BUG(strlen(bw_file_headers) > MAX_VOTE_LINE_LEN)) {
        tor_free(bw_file_headers);
        bw_file_headers = tor_strdup("");
      }
```

Reply:
Almost, but you can't overwrite bw_file_headers with valid data after an error:
```
      if (BUG(smartlist_len(v3_ns->bw_file_headers) > MAX_BW_FILE_HEADERS_LEN)) {
        bw_file_headers = tor_strdup("");
      } else {
        bw_file_headers = smartlist_join_strings(v3_ns->bw_file_headers, " ", 0,
                                                 NULL);
        if (BUG(strlen(bw_file_headers) > MAX_VOTE_LINE_LEN)) {
          tor_free(bw_file_headers);
          bw_file_headers = tor_strdup("");
        }
    }
```

Here's another alternative that might make the code simpler:
```
      char *bw_file_headers = NULL;
      …
      if (! BUG(smartlist_len(v3_ns->bw_file_headers) > MAX_BW_FILE_HEADERS_LEN)) {
        bw_file_headers = smartlist_join_strings(v3_ns->bw_file_headers, " ", 0,
                                                 NULL);
        if (BUG(strlen(bw_file_headers) > MAX_VOTE_LINE_LEN)) {
          /* Free and set to NULL, so the vote header line is empty */
          tor_free(bw_file_headers);
        }
      }
      …
      bw_file_headers ? bw_file_headers : "");
```

Then the final tor_free(bw_file_headers) works because free(NULL) does nothing.